pandas/core/_numba/executor.py

```diff
@@ -0,0 +1,66 @@
+from __future__ import annotations
+
+from typing import (
+    Any,
+    Callable,
+)
+
+import numpy as np
+
+from pandas._typing import Scalar
+from pandas.compat._optional import import_optional_dependency
+
+from pandas.core.util.numba_ import (
+    NUMBA_FUNC_CACHE,
+    get_jit_arguments,
+)
+
+
+def generate_shared_aggregator(
+    func: Callable[..., Scalar],
+    kwargs: dict[str, Any],
+    engine_kwargs: dict[str, bool] | None,
+    cache_key_str: str,
+):
+    """
+    Generate a Numba function that loops over the columns 2D object and applies
+    a 1D numba kernel over each column.
+
+    Parameters
+    ----------
+    func : function
+        aggregation function to be applied to each column
+    kwargs : dict
+        **kwargs to be passed into the function. Should be unused as not
+        supported by Numba
+    engine_kwargs : dict
+        dictionary of arguments to be passed into numba.jit
+    cache_key_str: str
+        string to access the compiled function of the form
+        <caller_type>_<aggregation_type> e.g. rolling_mean, groupby_mean
+
+    Returns
+    -------
+    Numba function
+    """
+    nopython, nogil, parallel = get_jit_arguments(engine_kwargs, kwargs)
+
+    cache_key = (func, cache_key_str)
+    if cache_key in NUMBA_FUNC_CACHE:
+        return NUMBA_FUNC_CACHE[cache_key]
+
+    numba = import_optional_dependency("numba")
+
+    @numba.jit(nopython=nopython, nogil=nogil, parallel=parallel)
+    def column_looper(
+        values: np.ndarray,
+        start: np.ndarray,
+        end: np.ndarray,
+        min_periods: int,
+    ):
+        result = np.empty((len(start), values.shape[1]))
+        for i in numba.prange(values.shape[1]):
```

Comment:
Do we have evidence of performance gains from using `prange` like this? From my (limited) experience with `prange`, it seems to only improve performance in very well structured cases where minimal operations happen in each looping step (like the reductions in https://numba.pydata.org/numba-doc/latest/user/parallel.html). I'm not sure if `prange` can ever cause a slowdown though, just curious if you have experience with it

Reply:
In the example above, looks like parallel performance is equivalent on my machine.

```
In [3]: %timeit roll.mean(engine="numba", engine_kwargs={"nopython": True, "nogil": True, "parallel": True})
283 ms ± 13 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [5]: %timeit roll.mean(engine="numba", engine_kwargs={"nopython": True, "nogil": True, "parallel": False})
284 ms ± 14.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

I have never seen it as a slowdown in other uses. `prange` is also an alas for `range` in `parallel=False` use cases. For this use case, the user is able to toggle parallel performance.